### ISSUE TITLE
NestedScopeFunctions false positive on Gradle's `Property<T>.set`

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5342]
## Steps to Reproduce
```kotlin
// Needs `implementation(gradleApi())` and `kotlin-dsl` plugin to compile.
import org.gradle.api.publish.maven.MavenPublication

fun MavenPublication.setupLinks(project: Project) {
	pom {
		url.set("https://github.com/detekt/detekt")
		scm {
			connection.set("scm:git:github.com/detekt/detekt.git") // 120
			developerConnection.set("scm:git:ssh://github.com/detekt/detekt.git") // 121
			url.set("https://github.com/detekt/detekt/tree/main") // 122
		}
		licenses {
			license { // 125
				name.set("Apache License 2.0") // 126
				url.set("https://github.com/detekt/detekt/blob/v${project.version}/LICENCE") // 127
			}
		}
	}
}
```

## Observed Behavior
```
...\PublishingPlugin.kt:120:15: The scope function 'set' is nested too deeply ('2'). Complexity threshold is set to '1'. [NestedScopeFunctions]
...\PublishingPlugin.kt:121:24: The scope function 'set' is nested too deeply ('2'). Complexity threshold is set to '1'. [NestedScopeFunctions]
...\PublishingPlugin.kt:122:8: The scope function 'set' is nested too deeply ('2'). Complexity threshold is set to '1'. [NestedScopeFunctions]
...\PublishingPlugin.kt:125:4: The scope function 'license' is nested too deeply ('2'). Complexity threshold is set to '1'. [NestedScopeFunctions]
...\PublishingPlugin.kt:126:10: The scope function 'set' is nested too deeply ('3'). Complexity threshold is set to '1'. [NestedScopeFunctions]
...\PublishingPlugin.kt:127:9: The scope function 'set' is nested too deeply ('3'). Complexity threshold is set to '1'. [NestedScopeFunctions]
```

## Expected Behavior
`set` should definitely not be flagged, as it's not a scoped function, it's just a function call on an object.
 
The `license` flag might be valid, although this is just how DSLs work in general. In this issue let's focus on `set`.

## Context
This code is in a module in my build logic in [PublishingPlugin.kt](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/d9e553b2e0896a7168e118284b6fd70a240987d4/gradle/plugins/src/main/kotlin/net/twisterrob/gradle/build/PublishingPlugin.kt#L116-L138).

## Your Environment

* Version of detekt used: 1.21.0
* Version of Gradle used (if applicable): 7.5.1
* Link to your project (if it's a public repository): https://github.com/TWiStErRob/net.twisterrob.gradle/blob/d9e553b2e0896a7168e118284b6fd70a240987d4/gradle/plugins/src/main/kotlin/net/twisterrob/gradle/build/PublishingPlugin.kt#L116-L138
* Running `detektMain` so type resolution is on.

**Comment by TWiStErRob** [comment]
A little bit of investigation:

<details><summary>When not using `kotlin-dsl` it doesn't report this problem</summary>

```kotlin
import org.gradle.api.publish.maven.MavenPublication

fun MavenPublication.setupLinks(project: Project) {
	pom {
		it.url.set("https://github.com/detekt/detekt")
		it.scm {
			it.connection.set("scm:git:github.com/detekt/detekt.git") // 120
			it.developerConnection.set("scm:git:ssh://github.com/detekt/detekt.git") // 121
			it.url.set("https://github.com/detekt/detekt/tree/main") // 122
		}
		it.licenses {
			it.license { // 125
				it.name.set("Apache License 2.0") // 126
				it.url.set("https://github.com/detekt/detekt/blob/v${project.version}/LICENCE") // 127
			}
		}
	}
}
```
</details>

<details><summary>When trying to reproduce in pure Kotlin, it doesn't report this problem</summary>

```kotlin
fun MavenPublication.setupLinks() {
	pom {
		url.set("https://github.com/detekt/detekt")
		scm {
			connection.set("scm:git:github.com/detekt/detekt.git") // 120
			developerConnection.set("scm:git:ssh://github.com/detekt/detekt.git") // 121
			url.set("https://github.com/detekt/detekt/tree/main") // 122
		}
		licenses {
			license { // 125
				name.set("Apache License 2.0") // 126
				url.set("https://github.com/detekt/detekt/blob/main/LICENCE") // 127
			}
		}
	}
}

interface Provider<T> {
	fun get(): T
	fun getOrNull(): T?
}

interface Property<T> : Provider<T> {
	fun set(value: T)
	fun set(provider: Provider<T>)
}

interface MavenPublication {
	val pom: MavenPom
	fun pom(configure: MavenPom.() -> Unit)
}

interface MavenPom {
	val url: Property<String>
	fun scm(action: MavenPomScm.() -> Unit)
	fun licenses(action: MavenPomLicenseSpec.() -> Unit)
}

interface MavenPomScm {
	val connection: Property<String>
	val developerConnection: Property<String>
	val url: Property<String>
}

interface MavenPomLicenseSpec {
	fun license(action: MavenPomLicense.() -> Unit)
}

interface MavenPomLicense {
	val name: Property<String>
	val url: Property<String>
}
```

</details>

<details><summary>When trying to reproduce with Java API, I laughably failed to realize there's no receiver there while doing Kotlin -> Java conversion.</summary>

```java
interface Provider<T> {
	T get();
	@Nullable T getOrNull();
}

interface Property<T> extends Provider<T> {
	void set(T value);
	void set(Provider<T> provider);
}

interface MavenPublication {
	MavenPom getPom();
	void pom(configure: MavenPom.() -> Unit)
}

interface MavenPom {
	Property<String> getUrl();
	void scm(action: MavenPomScm.() -> Unit)
	void licenses(action: MavenPomLicenseSpec.() -> Unit)
}

interface MavenPomScm {
	Property<String> getConnection();
	Property<String> getDeveloperConnection();
	Property<String> getUrl();
}

interface MavenPomLicenseSpec {
	void license(action: MavenPomLicense.() -> Unit)
}

interface MavenPomLicense {
	Property<String> getName();
	Property<String> getUrl();
}
```
</details>

**Comment by BraisGabin** [comment]
I think this is a black magic that the `kotlin-dsl` does that goes beyond what detekt should do. If a fix is provided in a PR will could review it but otherwise this is a "won't fix".

**Comment by TWiStErRob** [comment]
The black magic that kotlin-dsl does is apply 2 or 3 compiler plugins, so this is a general issue. Right now it's only because it transforms `Action<T> { it: T -> }` to `{ this: T -> }`, but there are plans to transform `property = value` to `property.set(value)` for simpler-looking lazy configuration in gradle.kts.

@3flex do you have any idea what would it take to apply compiler plugins along with type resolution, or if this is out of the question; maybe using the detekt compiler plugin would "see" those `kotlin-dsl` transformations?

**Comment by 3flex** [comment]
I think we should forget adding support for analysing code after a third party compiler plugin is applied in the standalone version of detekt. Possibly our compiler plugin will support it, but it may not be feasible - I have no idea :/
